server/src/web/broadcast.rs

```diff
@@ -0,0 +1,118 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+use chrono::Utc;
+use futures::future::join_all;
+use log::debug;
+use serene_data::package::BroadcastEvent;
+use tokio::sync::Mutex;
+use actix_web_lab::sse;
+use actix_web_lab::sse::Sse;
+use actix_web_lab::util::InfallibleStream;
+use tokio_stream::wrappers::ReceiverStream;
+
+pub enum Event {
+    BuildStart,
+    BuildFinish,
+    Log(String)
+}
+
+pub struct Broadcast {
+    subscriptions: Mutex<HashMap<String, Vec<tokio::sync::mpsc::Sender<sse::Event>>>>,
+    // cache contains build logs for packages which are currently building
+    cache: Mutex<HashMap<String, Vec<String>>>,
+}
+
+impl Broadcast {
+    pub fn new() -> Arc<Self> {
+        let broadcast = Arc::new(Self { 
+            subscriptions: Mutex::new(HashMap::new()),
+            cache: Mutex::new(HashMap::new()),
+        });
+        Broadcast::spawn_ping(broadcast.clone());
+        broadcast
+    }
+
+    fn spawn_ping(this: Arc<Self>) {
+        tokio::task::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(10));
+
+            loop {
+                interval.tick().await;
+                this.remove_stale_connections().await;
+            }
+        });
+    }
+
+    /// remove all connections for which the ping broadcast fails to avoid unnecessary broadcasts
+    async fn remove_stale_connections(&self) {
+        let mut subscriptions = self.subscriptions.lock().await;
+
+        *subscriptions = join_all(
+            subscriptions.iter().map(|(package, receivers)| async {
+                let receivers = join_all(
+                    receivers.iter().map(|recv| async {
+                        let event = sse::Event::Data(sse::Data::new(Utc::now().timestamp_millis().to_string()).event(BroadcastEvent::Ping.to_string()));
+                        recv.send(event).await.ok().map(|_| recv.clone())
+                    })
+                ).await.into_iter().filter_map(|r| r).collect::<Vec<_>>();
```

Comment:
style nit: there is `.flatten()`  for the use of filter map here, also use is_empty below